File: python_code/problem1_1/test/UniqChars_Parameterized_Test.py
# coding=utf8
from problem1_1.src.UniqChars import UniqChars
from parameterizedtestcase import ParameterizedTestCase


class UniqChars_Parameterized_Test(ParameterizedTestCase):
    @ParameterizedTestCase.parameterize(
        ("string", "method", "expected"),
        [
            ("", UniqChars.BIT, False),
            ("", UniqChars.BOOL, False),
            ("", UniqChars.SORTED, False),
            ("a", UniqChars.BIT, True),
            ("a", UniqChars.BOOL, True),
            ("a", UniqChars.SORTED, True),
            ("ab", UniqChars.BIT, True),
            ("ab", UniqChars.BOOL, True),
            ("ab", UniqChars.SORTED, True),
            ("aba", UniqChars.BIT, False),
            ("aba", UniqChars.BOOL, False),
            ("aba", UniqChars.SORTED, False),
            ("aaba", UniqChars.BIT, False),
            ("aaba", UniqChars.BOOL, False),
            ("aaba", UniqChars.SORTED, False),
            ("baa", UniqChars.BIT, False),
            ("baa", UniqChars.BOOL, False),
            ("baa", UniqChars.SORTED, False),
            ("abcdefghijklmnopqrstuvwxyz{|}~_`", UniqChars.BIT, True),
            ("abcdefghijklmnopqrstuvwxyz{|}~_`", UniqChars.BOOL, True),
            ("abcdefghijklmnopqrstuvwxyz{|}~_`", UniqChars.SORTED, True),
            (("abcdefghijklmnopqrstuvwxyz{|}~_`"
                "abcdefghijklmnopqrstuvwxyz{|}~_`"), UniqChars.BIT, False),
            (("abcdefghijklmnopqrstuvwxyz{|}~_`"
                "abcdefghijklmnopqrstuvwxyz{|}~_`"), UniqChars.BOOL, False),
            (("abcdefghijklmnopqrstuvwxyz{|}~_`"
                "abcdefghijklmnopqrstuvwxyz{|}~_`"), UniqChars.SORTED, False),
            ("!🛅", UniqChars.BIT, True),
            ("!🛅", UniqChars.BOOL, True),
            ("!🛅", UniqChars.SORTED, True),
            ("!🛅🛅", UniqChars.BIT, False),
            ("!🛅🛅", UniqChars.BOOL, False),
            ("!🛅🛅", UniqChars.SORTED, False),
        ]
    )
    def test_uniqChars_method(self, string, method, expected):
        self.assertEqual(UniqChars.uniqChars(string, method), expected)

File: java_code/src/test/java/com/artvb/iq/problem1_1/UniqCharsParameterizedTest.java
package com.artvb.iq.problem1_1;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.util.Arrays;

import static org.junit.Assert.assertEquals;

@RunWith(value = Parameterized.class)
public class UniqCharsParameterizedTest {

    private String str;
    private int method;
    private boolean expected;

    public UniqCharsParameterizedTest(String str, int method, boolean expected) {
        this.str = str;
        this.method = method;
        this.expected = expected;
    }

    @Parameterized.Parameters(name = "{index}: uniqChars({0}, {1}) => {2}")
    public static Iterable<Object[]> data() {
        return Arrays.asList(new Object[][]{
                {"", UniqChars.BIT, false},
                {"", UniqChars.BOOL, false},
                {"", UniqChars.SORTED, false},
                {"a", UniqChars.BIT, true},
                {"a", UniqChars.BOOL, true},
                {"a", UniqChars.SORTED, true},
                {"ab", UniqChars.BIT, true},
                {"ab", UniqChars.BOOL, true},
                {"ab", UniqChars.SORTED, true},
                {"aba", UniqChars.BIT, false},
                {"aba", UniqChars.BOOL, false},
                {"aba", UniqChars.SORTED, false},
                {"aaba", UniqChars.BIT, false},
                {"aaba", UniqChars.BOOL, false},
                {"aaba", UniqChars.SORTED, false},
                {"baa", UniqChars.BIT, false},
                {"baa", UniqChars.BOOL, false},
                {"baa", UniqChars.SORTED, false},
                {"abcdefghijklmnopqrstuvwxyz{|}~_`", UniqChars.BIT, true},
                {"abcdefghijklmnopqrstuvwxyz{|}~_`", UniqChars.BOOL, true},
                {"abcdefghijklmnopqrstuvwxyz{|}~_`", UniqChars.SORTED, true},
                {"abcdefghijklmnopqrstuvwxyz{|}~_`abcdefghijklmnopqrstuvwxyz{|}~_`", UniqChars.BIT, false},
                {"abcdefghijklmnopqrstuvwxyz{|}~_`abcdefghijklmnopqrstuvwxyz{|}~_`", UniqChars.BOOL, false},
                {"abcdefghijklmnopqrstuvwxyz{|}~_`abcdefghijklmnopqrstuvwxyz{|}~_`", UniqChars.SORTED, false},
                // The BIT method cannot be used with these tests because the char values are out of range
                {"!🛅", UniqChars.BOOL, true},
                {"!🛅", UniqChars.SORTED, true},
                {"!🛅🛅", UniqChars.BOOL, false},
                {"!🛅🛅", UniqChars.SORTED, false},
        });
    }

    @Test
    public void testUniqCharsFunction() {
        assertEquals(expected, UniqChars.uniqChars(str, method));
    }
}